RHIDP-1167: Software template for creating a frontend plugin (#5)
redhat-developer/red-hat-developer-hub-software-templates: templates/create-frontend-plugin/docs/features/api.md

## Guide to consuming APIs in your UI components

> The ExampleFetchComponent demonstrates the common task of making an asynchronous request to a public API and displaying the response data in a table using Material UI components.

Reviewer: ```suggestion
The ExampleFetchComponent demonstrates the common task of making an asynchronous request to a public API and displaying the response data in a table using Backstage components.
```

You can modify these components, rename them, or replace them entirely.

To consume APIs from your backend plugin, follow these steps:

- Create an `api` folder under your `src` directory and add a client file for making API calls. The file name should follow the format `<PluginNameInTitleCase>BackendClient.ts`.

    The file structure should be as follows:

>     ```ts title="<PluginNameInTitleCase>BackendClient.ts"

Reviewer: Does &lg; &gt; works here? Because the < and > isn't rendered:

![image](https://github.com/user-attachments/assets/1f00481c-9e9b-4b0c-a80e-be110de7a5c0)

        import {
        ConfigApi,
        createApiRef,
        IdentityApi,
        } from '@backstage/core-plugin-api';

        // @public
        export type <PluginNameInTitleCase>API = {
        getUsers: () => Promise<{ status: string } | Response>;

>         };

Reviewer: The complete code should be intended correctly.

        export type Options = {
        configApi: ConfigApi;
        identityApi: IdentityApi;
        };

        // @public
        export const <PluginNameInCamelCase>ApiRef = createApiRef<<PluginNameInTitleCase>API>({
        id: 'plugin.<PluginNameInCamelCase>.service',
        });

        export class <PluginNameInTitleCase>BackendClient implements <PluginNameInTitleCase>API {

        private readonly configApi: ConfigApi;
        private readonly identityApi: IdentityApi;

        constructor(options: Options) {
            this.configApi = options.configApi;
            this.identityApi = options.identityApi;
        }

        async getUsers() {
            const { token: idToken } = await this.identityApi.getCredentials();
            const backendUrl = this.configApi.getString('backend.baseUrl');
            const jsonResponse = await fetch(`${backendUrl}/api/<mention-your-api-here>/`, {
            headers: {
                ...(idToken && { Authorization: `Bearer ${idToken}` }),
            },
            });
            return jsonResponse.json();
        }
        }
    ```

- After creating the client, consume the APIs in your components:

>     ```tsx title="ExampleFetchComponent.ts"

Reviewer: Looks like tsx isn't supported:

![image](https://github.com/user-attachments/assets/285a0cc0-614b-4ff1-83f1-1533531df0c8)

```suggestion
    ```ts title="ExampleFetchComponent.ts"
```

        export const ExampleFetchComponent = () => {

>         const theme = useTheme();

Reviewer: intention

        const chipStyle = getChipStyle(theme);
        const pluginApi = useApi(<PluginNameInCamelCase>ApiRef); // use the api ref created in the client file
        const users = pluginApi.getUsers();
        const columns: TableColumn[] = [
            { title: 'Column title 1', field: 'column-field-1' },
            { title: 'Column title 2', field: 'column-field-2' },
            { title: 'Column title 3', field: 'column-field-3' },
        ];

        const data = users.map((user: any) => {
            // prepare your table data here
        });

        return (
            <Table
            title="<table-title>"
            options={{ search: false, paging: true }}
            columns={columns}
            data={data}
            />
        );
        };

    ```

You may customize the components as needed to fit your requirements.

## References

> RBAC front-end plugin: https://github.com/janus-idp/backstage-plugins/tree/main/plugins/rbac

Reviewer: ```suggestion
RBAC frontend plugin: https://github.com/janus-idp/backstage-plugins/tree/main/plugins/rbac
```

> Orchestrator front-end plugin: https://github.com/janus-idp/backstage-plugins/tree/main/plugins/orchestrator

Reviewer: ```suggestion
Orchestrator frontend plugin: https://github.com/janus-idp/backstage-plugins/tree/main/plugins/orchestrator
```
